File: src/main/java/org/lanternpowered/server/status/LanternStatusResponse.java
/*
 * This file is part of LanternServer, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://github.com/LanternPowered>
 * Copyright (c) Contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.server.status;

import static com.google.common.base.Preconditions.checkNotNull;

import org.spongepowered.api.MinecraftVersion;
import org.spongepowered.api.event.server.ClientPingServerEvent;
import org.spongepowered.api.network.status.Favicon;
import org.spongepowered.api.text.Text;

import java.util.Optional;

import javax.annotation.Nullable;

public class LanternStatusResponse implements ClientPingServerEvent.Response {

    private final MinecraftVersion version;

    @Nullable private Favicon favicon;
    private Text description;
    private Players players;

    private boolean hidePlayers;

    public LanternStatusResponse(MinecraftVersion version, Favicon favicon, Text description, Players players) {
        this.description = description;
        this.favicon = favicon;
        this.version = version;
        this.players = players;
    }

    @Override
    public Text getDescription() {
        return this.description;
    }

    @Override
    public MinecraftVersion getVersion() {
        return this.version;
    }

    @Override
    public Optional<Favicon> getFavicon() {
        return Optional.ofNullable(this.favicon);
    }

    @Override
    public void setDescription(Text description) {
        this.description = checkNotNull(description, "description");
    }

    @Override
    public void setHidePlayers(boolean hide) {
        this.hidePlayers = hide;
    }

    @Override
    public void setFavicon(@Nullable Favicon favicon) {
        this.favicon = favicon;
    }

    @Override
    public Optional<Players> getPlayers() {
        if (this.hidePlayers) {
            return Optional.empty();
        }
        return Optional.of(this.players);
    }

    /**
     * Gets the instance of the players safely, so it will never be null.
     * 
     * @return the players
     */
    public Players getPlayerSafely() {
        return this.players;
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/status/LanternStatusResponse.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.status

import org.lanternpowered.api.MinecraftVersion
import org.lanternpowered.api.text.Text
import org.lanternpowered.api.util.ToStringHelper
import org.lanternpowered.api.util.optional.emptyOptional
import org.lanternpowered.api.util.optional.asOptional
import org.lanternpowered.server.game.version.LanternMinecraftVersion
import org.spongepowered.api.event.server.ClientPingServerEvent
import org.spongepowered.api.network.status.Favicon

class LanternStatusResponse(
        private var version: MinecraftVersion,
        private var description: Text,
        private val players: ClientPingServerEvent.Response.Players,
        private var favicon: Favicon?,
        private var hidePlayers: Boolean = false
) : ClientPingServerEvent.Response {


    /**
     * Sets the [MinecraftVersion] of the status response.
     *
     * The name of this version will be displayed on the client when the
     * server or the client is outdated.
     *
     * @param version The version
     */
    fun setVersion(version: MinecraftVersion) { this.version = version }

    /**
     * Sets the [MinecraftVersion] of the status response.
     *
     * The name of this version will be displayed on the client when the
     * server or the client is outdated.
     *
     * @param name The name
     * @param protocol The protocol version
     * @param legacy Whether the version is legacy
     */
    fun setVersion(name: String, protocol: Int, legacy: Boolean) {
        setVersion(LanternMinecraftVersion(name, protocol, legacy))
    }

    override fun getDescription() = this.description
    override fun getVersion() = this.version
    override fun getFavicon() = this.favicon.asOptional()
    override fun getPlayers() = if (this.hidePlayers) emptyOptional() else this.players.asOptional()

    override fun setDescription(description: Text) = run { this.description = description }
    override fun setHidePlayers(hide: Boolean) = run { this.hidePlayers = hide }
    override fun setFavicon(favicon: Favicon?) = run { this.favicon = favicon }

    override fun toString(): String = ToStringHelper(this)
            .omitNullValues()
            .add("version", this.version)
            .add("description", this.description)
            .add("players", this.players)
            .add("hidePlayers", this.hidePlayers)
            .add("favicon", this.favicon)
            .toString()
}
